pennylane/_qubit_device.py

```diff
@@ -29,100 +30,155 @@
-        self._memory = None
+        self._wires_measured = set()
         self._samples = None
 
-    def execute(self, queue, observables, parameters=None):
-        """Execute a queue of quantum operations on the device and then measure the given observables.
+    def execute(self, circuit):
+        """Execute a queue of quantum operations on the device and then
+        measure the given observables.
+
+        For plugin developers: instead of overwriting this, consider
+        implementing a suitable subset of
+
+        * :meth:`apply`
 
-        For plugin developers: Instead of overwriting this, consider implementing a suitable subset of
-        :meth:`pre_apply`, :meth:`apply`, :meth:`post_apply`, :meth:`pre_measure`,
-        :meth:`expval`, :meth:`var`, :meth:`sample`, :meth:`post_measure`, and :meth:`execution_context`.
+        * :meth:`~.generate_samples`
+
+        * :meth:`~.probability`
 
         Args:
-            queue (Iterable[~.operation.Operation]): operations to execute on the device
-            observables (Iterable[~.operation.Observable]): observables to measure and return
-            parameters (dict[int->list[ParameterDependency]]): Mapping from free parameter index to the list of
-                :class:`Operations <pennylane.operation.Operation>` (in the queue) that depend on it.
+            circuit (~.CircuitGraph): circuit to execute on the device
 
         Raises:
             QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
 
         Returns:
             array[float]: measured value(s)
         """
-        if parameters is None:
-            parameters = {}
+        self.check_validity(circuit.operations, circuit.observables)
 
-        self.check_validity(queue, observables)
-        self._op_queue = queue
-        self._obs_queue = observables
-        self._parameters = {}
-        self._parameters.update(parameters)
+        # apply all circuit operations
+        self.apply(circuit.operations, circuit.diagonalizing_gates)
 
-        with self.execution_context():
-            self.pre_apply()
+        # determine the wires that are measured by the circuit
+        self._wires_measured = QubitDevice.active_wires(circuit.observables)
 
-            for operation in queue:
-                # Pass instances directly
-                self.apply(operation)
+        # generate computational basis samples
+        if (not self.analytic) or circuit.is_sampled:
+            self.generate_samples()
 
-            self.rotate_basis(observables)
+        # compute the required statistics
+        results = self.statistics(circuit.observables)
 
-            self.post_apply()
+        # Ensures that a combination with sample does not put
+        # expvals and vars in superfluous arrays
+        sample_return_types = (obs.return_type is Sample for obs in circuit.observables)
+        if any(sample_return_types) and not all(sample_return_types):
```

Comment:
This part could now be refactored such that it uses ``circuit.is_sampled``. Not sure if a ``circuit.only_sampling`` property would be needed (for the second boolean), but might be the case (or having a comparable name for ``all(obs.return_type is Sample for obs in circuit.observables)``.

Reply:
Done!